cmd/maestro/server/pulse_server.go

```diff
@@ -5,130 +5,173 @@ import (
-				logger.Error(fmt.Sprintf("Unable to update heartbeat for maestro instance: %s", err.Error()))
-			}
-		case <-checkTicker.C:
-			logger.V(4).Infof("Checking maestro instances liveness and trigger statusresync for dead instances")
-			if err := s.CheckInstances(ctx); err != nil {
-				// log and ignore the error and continue to tolerate the intermittent issue
-				logger.Error(fmt.Sprintf("Unable to check maestro instances: %s", err.Error()))
-			}
-		}
+		log.Error(fmt.Sprintf("Unable to upsert maestro instance: %s", err.Error()))
 	}
 }
 
-func (s *PulseServer) CheckInstances(ctx context.Context) error {
+func (s *PulseServer) checkInstances() {
+	log.V(4).Infof("Checking liveness of maestro instances")
+	ctx := context.TODO()
 	// lock the Instance with a fail-fast advisory lock context.
 	// this allows concurrent processing of many instances by one or more maestro instances exclusively.
 	lockOwnerID, acquired, err := s.lockFactory.NewNonBlockingLock(ctx, "maestro-instances-pulse-check", db.Instances)
 	// Ensure that the transaction related to this lock always end.
 	defer s.lockFactory.Unlock(ctx, lockOwnerID)
 	if err != nil {
-		return fmt.Errorf("error obtaining the event lock: %v", err)
+		log.Error(fmt.Sprintf("error obtaining the instance lock: %v", err))
+		return
 	}
 	// skip if the lock is not acquired
 	if !acquired {
-		return fmt.Errorf("failed to acquire the lock as another maestro instance is checking instances")
+		log.Error(fmt.Sprintf("failed to acquire the lock as another maestro instance is checking instances"))
+		return
 	}
-	// Instances not pulsing within the last three check intervals are considered as dead.
-	instances, err := s.instanceDao.FindByUpdatedTime(ctx, time.Now().Add(time.Duration(int64(-3*time.Second)*s.checkInterval)))
+
+	instances, err := s.instanceDao.All(ctx)
 	if err != nil {
-		return fmt.Errorf("unable to get outdated maestro instances: %s", err.Error())
-	}
-	deletedInstanceIDs := []string{}
-	for _, i := range instances {
-		deletedInstanceIDs = append(deletedInstanceIDs, i.ID)
+		log.Error(fmt.Sprintf("Unable to get all maestro instances: %s", err.Error()))
+		return
 	}
 
-	if len(deletedInstanceIDs) > 0 {
-		// trigger statusresync for dead instances only once even if there are multiple dead instances
-		// will retry in next check if the statusresync fails
-
-		// send resync request to each consumer
-		// TODO: optimize this to only resync resource status for necessary consumers
-		consumerIDs := []string{}
-
-		consumers, err := s.consumerDao.All(ctx)
-		if err != nil {
-			return fmt.Errorf("unable to trigger statusresync for maestro instance(s): %s", err.Error())
+	inactiveInstanceIDs := []string{}
+	for _, instance := range instances {
+		// Instances pulsing within the last two check intervals are considered as active.
+		if instance.UpdatedAt.After(time.Now().Add(time.Duration(int64(-2*time.Second) * s.pulseInterval))) {
```

Comment:
nit, 2 seems to be too short to me. Given this is a poll, and if one pulse failed we do not retry immeditaly.

Reply:
changed to 3, not sure if 3 is reasonable value.